pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -304,20 +307,43 @@ func (e *endpointSliceCache) Update(hostname host.Name, slice string, endpoints
 	e.mu.Lock()
 	defer e.mu.Unlock()
 	if len(endpoints) == 0 {
-		delete(e.endpointsByServiceAndSlice[hostname], slice)
+		for _, ip := range e.endpointIPsByServiceAndSlice[hostname][slice] {
+			delete(e.endpointByIP, ip)
+		}
+		delete(e.endpointIPsByServiceAndSlice[hostname], slice)
+	}
+	if _, f := e.endpointIPsByServiceAndSlice[hostname]; !f {
+		e.endpointIPsByServiceAndSlice[hostname] = make(map[string][]string)
 	}
-	if _, f := e.endpointsByServiceAndSlice[hostname]; !f {
-		e.endpointsByServiceAndSlice[hostname] = make(map[string][]*model.IstioEndpoint)
+	ips := make([]string, 0, len(endpoints))
+	for _, ep := range endpoints {
+		ips = append(ips, ep.Address)
+		// We will always overwrite. A conflict here means an endpoint is transitioning
+		// from one slice to another See
+		// https://github.com/kubernetes/website/blob/master/content/en/docs/concepts/services-networking/endpoint-slices.md#duplicate-endpoints
+		// In this case, we can always assume and update is fresh, although older slices
+		// we have not gotten updates may be stale; therefor we always take the new
+		// update.
+		e.endpointByIP[ep.Address] = ep
```

Comment:
I think we need to delete `endpointIPsByServiceAndSlice[hostname][slice]` and add again like L329

And also need to delete all `endpointByIP` by iterating the existing slice endpoints ips, otherwise this is not correct in case only one endpoint deleted from a EPSlice.

Reply:
added delete. thanks